### PR TITLE
Remove YEAR field of all generated files and fix kubernetes boilerplate checker

### LINKED ISSUE #24: Copyright boilerplate changes YEAR, but shouldn't
The copyrighted code is either copyright the first year it was generated or else it is copyright the year we wrote gengo.  It obviously doesn't change copyright by nature of being regenerated.

### PR DESCRIPTION
**Pull request by fisherxu**
**What this PR does / why we need it**:
Remove YEAR field of all generated files and fix kubernetes boilerplate checker
xref: [remove YEAR fileds in gengo #91](https://github.com/kubernetes/gengo/pull/91)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes [#gengo/issues/24](https://github.com/kubernetes/gengo/issues/24)

**Special notes for your reviewer**:
/cc @thockin @lavalamp @sttts 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by cblecker** [comment]
@fisherxu Can you replace the release note with "NONE"? No release note is needed here.

**Comment by fisherxu** [comment]
@cblecker Sure, have done

**Comment by sttts** [comment]
/hold

This conflicts with https://github.com/kubernetes/kubernetes/pull/57656. We don't want the boilerplate without the year in kube, but it should come from the code-generator repo.

**Review by thockin — COMMENTED** [review]
I really dislike the idea of coding in a list of directories.  I don't have a better answer off the top of my head, but consider this a plea - can anyone please find a better approach to this?

@lavalamp

File: hack/boilerplate/boilerplate.py [hunk 1]
```diff
@@ -183,14 +197,84 @@ def get_regexs():
     regexs["shebang"] = re.compile(r"^(#!.*\n)\n*", re.MULTILINE)
     return regexs
 
+def generated_files(filename):
```
[the remark this answers is not in the document]
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: Have added some comments :) I also have no better approach to identify the generated files :(

File: hack/boilerplate/boilerplate.py [hunk 2]
```diff
@@ -133,6 +133,20 @@ def file_extension(filename):
                 "vendor", "test/e2e/generated/bindata.go", "hack/boilerplate/test",
                 "pkg/generated/bindata.go"]
 
+# list all the dirs that contain generated files
+generated_dirs = ['pkg/client/listers', 'pkg/client/clientset_generated', 'pkg/client/informers', 'k8s.io/sample-controller/pkg/client/listers',
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: I am also against having another directory list. Note that we will also have this "DO NOT EDIT" line in generated files. Can't we just use that to detect generated files?

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: yeah, it's a nice idea, but the comment in different files is also different, like [pkg/client/listers/admissionregistration/internalversion/expansion_generated.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/listers/admissionregistration/internalversion/expansion_generated.go)  and  [zz_generated.deepcopy.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/abac/zz_generated.deepcopy.go)

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: How about make the comments all the same? then may be we can identify them easily

Comment (sttts) [reply hack/boilerplate/boilerplate.py]: With an updated gengo it will be uniform.

Comment (sttts) [reply hack/boilerplate/boilerplate.py]: https://github.com/kubernetes/gengo/pull/94

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: OK, some files generated from script in hack maybe also need fixed.

**Comment by thockin** [comment]
Assign back to me once it is ready for approval.

**Comment by sttts** [comment]
Need rebase.

**Comment by fisherxu** [comment]
@sttts yeah, after [#59674](https://github.com/kubernetes/kubernetes/pull/59674) merged, will rebase and refix the checker.

**Comment by sttts** [comment]
@fisherxu #59674 merged

**Comment by fisherxu** [comment]
@sttts ok, will rebase and fix the checker :-)

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/59172)

**Comment by fisherxu** [comment]
@thockin @lavalamp @sttts PR has rebased and CI has green, it's ready for review now :-)

File: hack/boilerplate/boilerplate.generatego.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/*
```
Comment (sttts) [inline hack/boilerplate/boilerplate.generatego.txt]: generatego is not a term used elsewhere. Better maybe: generated_files or code-generator.

File: hack/boilerplate/boilerplate.py [hunk 4]
```diff
@@ -133,6 +133,10 @@ def file_extension(filename):
                 "vendor", "test/e2e/generated/bindata.go", "hack/boilerplate/test",
                 "pkg/generated/bindata.go"]
 
+# list all the files that are ungenerated
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: was is "ungenerated"?

File: hack/boilerplate/boilerplate.py [hunk 5]
```diff
@@ -181,16 +185,95 @@ def get_regexs():
...
+        print("Unable to open %s: %s" % (filename, exc), file=verbose_out)
+        return False
+
+    data = f.read()
+    f.close()
+
+    p = regexs["generated"]
+    found = p.search(data)
+
+    if found:
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: return found

File: hack/boilerplate/boilerplate.py [hunk 6]
```diff
@@ -181,16 +185,95 @@ def get_regexs():
...
+        return True
+    else:
+        return False
+
+
+def generated_file_passes(filename, refs, regexs):
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: this needs docs. It's not clear what this func does.

File: hack/boilerplate/boilerplate.py [hunk 7]
```diff
@@ -181,16 +185,95 @@ def get_regexs():
...
+
+    # remove build tags from the top of Go files
+    p = regexs["go_build_constraints"]
+    (data, found) = p.subn("", data, 1)
+
+    data = data.splitlines()
+
+    # if our test file is smaller than the reference it surely fails!
+    if len(ref) > len(data):
+        print('File %s smaller than reference (%d < %d)' %
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: does this really need a print?

File: hack/boilerplate/boilerplate.py [hunk 8]
```diff
@@ -181,16 +185,95 @@ def get_regexs():
...
     regexs = get_regexs()
     refs = get_refs()
     filenames = get_files(refs.keys())
 
     for filename in filenames:
-        if not file_passes(filename, refs, regexs):
-            print(filename, file=sys.stdout)
+        # if the file is automatically generated, then don't
+        # check the YEAR field.
+        if is_generated_files(filename, regexs):
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: why is `files` plural?

**Comment by sttts** [comment]
Please squash the changes into fewer commits.

File: hack/boilerplate/boilerplate.generatego.txt (hunk 3 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.generatego.txt]: How about boilerplate.generated_files.txt? and it only used to generated go files.

File: hack/boilerplate/boilerplate.py (hunk 7 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: It's just a copy from func #[file_passes](https://github.com/kubernetes/kubernetes/blob/master/hack/boilerplate/boilerplate.py#L96)

File: hack/boilerplate/boilerplate.py (hunk 6 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: OK

File: hack/boilerplate/boilerplate.py (hunk 7 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: yeah, it has some duplicated code with #file_passes, but also has some difference  in the middle. :(
>does this really need a print?

in the func print, has flag verbose like: print("Unable to open %s: %s" % (filename, exc), file=verbose_out) 
just used for debug, we can run like:
~#: python hack/boilerplate/boilerplate.py -v
then will print the debug log.

File: hack/boilerplate/boilerplate.py (hunk 5 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: Have addressed

File: hack/boilerplate/boilerplate.py (hunk 8 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: Have addressed

File: hack/boilerplate/boilerplate.py (hunk 7 above)
Comment (sttts) [reply hack/boilerplate/boilerplate.py]: I think it's fine. It looked like leftover debug output, but it isn't.

File: hack/boilerplate/boilerplate.py (hunk 4 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: Have addressed the comment.

**Comment by fisherxu** [comment]
@sttts After the code view is complete and addressed all the commets, I will squash  the commits. Maybe more easy to review in separated commits :-)

**Comment by thockin** [comment]
I scanned it - it seems OK but I don't have time for an in-depth review of python.  Does someone else?

**Comment by sttts** [comment]
> I scanned it - it seems OK but I don't have time for an in-depth review of python this week. Does someone else?

/cc @cblecker

File: hack/boilerplate/boilerplate.py (hunk 6 above)
Comment (sttts) [reply hack/boilerplate/boilerplate.py]: What is refs? What are the regexs? What is returned? What are the side-effects?

File: hack/boilerplate/boilerplate.py [hunk 9]
```diff
@@ -181,16 +185,91 @@ def get_regexs():
...
+        print('File %s smaller than reference (%d < %d)' %
+              (filename, len(data), len(ref)),
+              file=verbose_out)
+        return False
+
+    # trim our file to the same number of lines as the reference file
+    data = data[:len(ref)]
+
+    p = regexs["year"]
+    for d in data:
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: isn't this the only actual difference to file_passes plus line 224 above? Can't we merge the two functions and react differently depending on whether is_generated_file return true or false? Also we would have read the file already, so we can avoid reading it twice this way.

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: Yes, it has some differences in line 224 above, and also the debug log in line 246.
I have also considered merging them through the generate flag, but I'm afraid the code is too confusing. So separated them in two function.

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: @sttts  and I can try to merge them now

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: @sttts have merged and also squash change in 6 commits, thanks for review!

File: hack/boilerplate/boilerplate.py [hunk 10]
```diff
@@ -73,14 +94,18 @@ def file_passes(filename, refs, regexs):
     f.close()
 
     basename = os.path.basename(filename)
-    extension = file_extension(filename)
+    if generated:
```
Comment (sttts) [inline hack/boilerplate/boilerplate.py]: we have the file read into data here already. Shouldn't we do the twoliner of line 79+80 here?

**Comment by sttts** [comment]
Some nit. Otherwise lgtm.

File: hack/boilerplate/boilerplate.py (hunk 10 above)
Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: At beginning of func is_generated_file, there is also a judgement to the script that contain "DO NOT EDIT".
I think all the judgement of is_generated_file in one func is more clear.

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: how about pass the data to is_generated_file, avoid read it twice？

Comment (fisherxu) [reply hack/boilerplate/boilerplate.py]: @sttts Have done, PTAL

**Comment by sttts** [comment]
Please squash.

**Comment by fisherxu** [comment]
@sttts Have squashed to 3 commits, PTAL

File: hack/.golint_failures [hunk 11]
```diff
@@ -449,12 +449,6 @@ staging/src/k8s.io/api/storage/v1alpha1
 staging/src/k8s.io/api/storage/v1beta1
 staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr
 staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis/cr/v1
-staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned
```
Comment (thockin) [inline hack/.golint_failures]: This seems unrelated to the rest of the PR?

**Comment by fisherxu** [comment]
Thank you all for review and should we pick it to the 1.10 branch along with #60866 ? @thockin @thockin @cblecker

File: hack/.golint_failures (hunk 11 above)
Comment (fisherxu) [reply hack/.golint_failures]: Run hack/update-all.sh don't contain file in the apiextensions-apiserver/examples, so I run the script in apiextensions-apiserver/examples manually.
